blog/_posts/2017-11-28-view-based-collections.md

```diff
@@ -0,0 +1,169 @@
+Creating this intermediate collection can be an expensive operation that we can
+avoid by using a view, which is a lazy collection: `xs.view.take(n).map(f)`.
+
+## What’s Wrong With Views?
+
+Unfortunately, the current support of views in the standard library suffers from
+two issues.
+
+### Operations Default Implementations Break Laziness
+
+In the current collections the default implementations of operations that return
+new collections (e.g. `map`, `filter`, `++`) are based on strict builders.
+Consequently, when these operations are inherited by lazy collections they break
+their laziness. Therefore, these operations have to carefully be overridden by lazy
+collections to disable the default behaviour and preserve laziness instead.
+From a maintainer point of view, this is a headache.
+
+### `CanBuildFrom` Doesn’t Work With Views
+
+Unfortunately, views not only cause trouble for the standard library maintainers
+but also end users. Indeed, the main mechanism to extend the collections
+with new operations, namely `CanBuildFrom`, is incompatible with views. This means
+that users that use an implicit `CanBuildFrom` instance to build a collection will
+get a runtime exception if that instance has been resolved for a view.
+This is because `CanBuildFrom`, as its name suggests, also uses a strict
+“builder based” approach. Therefore no sensible implementation of its methods can
+be provided for lazy collections.
+
+## Summary
+
+We have seen that on one hand views are useful to avoid unnecessary computations but
+the other hand they give pain to maintainers and are dangerous for end users.
+
+The collections redesign is an opportunity to fix these issues, as
+explained in the remainder of this article.
+
+## “View Based” Default Implementations
+
+Essentially, in the new collections we reversed the situation.
+Instead of having strict by default operations, they are now lazy by default and
+concrete collection types decide whether to evaluate their result or not. For instance,
+here is the implementation of `map`:
+
+~~~ scala
+trait IterableOps[A, CC[_], C] {
+  def map[B](f: A => B): CC[B] = fromIterable(View.Map(this, f))
+}
+~~~
+
+(where `CC[_]` is `List` or `Vector`)
+
+We can break down this implementation into two steps.
+
+First, we create a lazy `View` that “records” that the `map` operation has been applied.
+Constructing such a view is a cheap operation, here is the implementation of
+`View.Map`:
+
+~~~ scala
+object View {
+  case class Map[A, B](underlying: Iterable[A], f: A => B) extends View[B] {
+    def iterator() = underlying.iterator().map(f)
+  }
+}
+~~~
+
+As you can see, unless we actually iterate on the view we don’t evaluate neither the elements
+of the `underlying` collection nor the mapped elements.
+
+Second, we build the resulting collection from the view by calling `fromIterable`. This operation
+is abstract in `IterableOps` and is implemented by leaf classes of the hierarchy (e.g.
```

Comment:
e.g.,

Reply:
omitting the comma is allowed in UK english.